chainstate/types/src/block_index.rs

```diff
@@ -83,6 +86,14 @@ impl BlockIndex {
         &self.some_ancestor
     }
 
+    pub fn status(&self) -> BlockStatus {
+        self.status
+    }
+
+    pub fn set_status(&mut self, status: BlockStatus) {
```

Comment:
Perhaps instead of freely setting the status here, you can implement the stepping here too, somehow, and do a read-modify-write operation such that the status cannot be freely changed to a completely different object.

Reply:
IMO it will overcomplicate things. First of all, in `perform_block_checks` (AKA `integrate_block`) I'll have to have a `&mut BlockIndex` instead of just `&mut BlockStatus`, so more space for errors. Secondly, `perform_block_checks` needs to persist a `BlockIndex` before calling `activate_best_chain`; right now it persists one with the "Initial" status, but after the proposed change it will have to persist `BlockIndex` with some intermediate status, which doesn't feel nice.
So unless you insist, I'd leave it as is. (Or maybe I don't understand your idea fully)

Reply:
[discussed] I'll skip this for now.